### PR TITLE
millet 0.9.4 (new formula)

### PR DESCRIPTION
**Pull request by arturovm**
Millet is an LSP server for Standard ML, written in Rust

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/millet.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
+require "json"
+
+class Millet < Formula
+  desc "Language server for Standard ML (SML)"
+  homepage "https://github.com/azdavis/millet"
+  url "https://github.com/azdavis/millet.git",
+      tag:      "v0.9.4",
+      revision: "f99f096471887deae67e3569b30df07aa7b33acb"
```
Comment (chenrui333) [inline Formula/millet.rb:8]: why not use `https://github.com/azdavis/millet/archive/refs/tags/v0.9.4.tar.gz`?

File: Formula/millet.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  head "https://github.com/azdavis/millet.git", branch: "main"
+
+  depends_on "rust" => :build
+
+  def install
+    system "cargo", "install", *std_cargo_args(path: "crates/millet-ls")
+  end
+
+  test do
+    initialize_request = { jsonrpc: "2.0",
```
Comment (chenrui333) [inline Formula/millet.rb:19]: there are many LSP examples in the repo, maybe you can just adopt the tests from the others?

File: Formula/millet.rb (hunk 1 above)
Comment (arturovm) [reply Formula/millet.rb:8]: @chenrui333 That's what I had at first, but seeing how everything is based on git around here, I thought maybe this was preferable. Is it preferable then to link to the tarball?

File: Formula/millet.rb (hunk 2 above)
Comment (arturovm) [reply Formula/millet.rb:19]: @chenrui333 I tried with one example, but this particular LSP server errors out when `stdin` is closed without the proper shutdown sequence, so it felt icky to me to write my tests around that. Wdyt?

Btw, thanks for your review 👍🏽

Comment (arturovm) [reply Formula/millet.rb:19]: @chenrui333 Also, for some reason using `pipe_output` results in a flaky test that sometimes passes and sometimes not. I think it's related to the `sync` flag on `pipe` and this server being weird.

File: Formula/millet.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
+require "json"
+
```
Comment (carlocab) [inline Formula/millet.rb:2]:
```suggestion
```
You don't need this.

**Review by carlocab — COMMENTED** [review]
Thanks; one suggestion. Please squash all your commits together when you apply this.

**Comment by arturovm** [comment]
> Thanks; one suggestion. Please squash all your commits together when you apply this.

@carlocab Thanks! Is "squash and merge" enabled for this repo? Or do I have to do that manually on my branch?

File: Formula/millet.rb (hunk 1 above)
Comment (chenrui333) [reply Formula/millet.rb]: yeah, that is right, we'd like to use tarball whenever possible

**Comment by chenrui333** [comment]
yeah, it would be good to squash on your end :)

**Comment by carlocab** [comment]
> Is "squash and merge" enabled for this repo?

It is not, because it doesn't work for our merge flow.

> Or do I have to do that manually on my branch?

Yes, else I would not have asked.

**Comment by arturovm** [comment]
@chenrui333 @carlocab Done! Thank you both for your reviews, really appreciate it.

**Review by carlocab — DISMISSED** [review]
Thanks!

File: Formula/millet.rb (hunk 2 above)
Comment (chenrui333) [reply Formula/millet.rb:16]: got you!

**Review by chenrui333 — DISMISSED** [review]

**Pull request merged**